pkg/provider/identity.go

```diff
@@ -0,0 +1,111 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"regexp"
+
+	azaci "github.com/Azure/azure-sdk-for-go/services/containerinstance/mgmt/2021-10-01/containerinstance"
+	client2 "github.com/virtual-kubelet/azure-aci/pkg/client"
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+	armcontainerservice "github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/containerservice/armcontainerservice"
+	armmsi "github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/msi/armmsi"
+	azidentity "github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+)
+
+func SetContainerGroupIdentity(identity *armmsi.Identity, identityType azaci.ResourceIdentityType, containerGroup *client2.ContainerGroupWrapper) {
+	if identity == nil || identityType != azaci.ResourceIdentityTypeUserAssigned {
+		return
+	}
+
+	cgIdentity := azaci.ContainerGroupIdentity{
+		Type: identityType,
+		UserAssignedIdentities: map[string]*azaci.ContainerGroupIdentityUserAssignedIdentitiesValue{
+			*identity.ID: &azaci.ContainerGroupIdentityUserAssignedIdentitiesValue{
+				PrincipalID: identity.Properties.PrincipalID,
+				ClientID: identity.Properties.ClientID,
+			},
+		},
+	}
+	containerGroup.Identity = &cgIdentity
+}
+
+func GetAgentPoolKubeletIdentity(ctx context.Context, providerResourceGroup string, providerVnetSubscriptionID string) (*armmsi.Identity, error) {
+
+	// initialize msi  credentials move this to setup
+	cred, err := azidentity.NewDefaultAzureCredential(nil)
+	if err != nil {
+		return nil, err
+	}
+	client, err := armmsi.NewUserAssignedIdentitiesClient(providerVnetSubscriptionID, cred, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// default MC_ resource group
+	if strings.HasPrefix(providerResourceGroup, "MC_") {
+		// use sdk to list identities by RG
+		pager := client.NewListByResourceGroupPager(providerResourceGroup, nil)
+		for pager.More() {
+			// pick the agent pool identity
+			nextResult, err := pager.NextPage(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, v := range nextResult.Value {
+				if strings.HasSuffix(*v.ID, "agentpool") {
+					return v, nil
+				}
+			}
+		}
+	}
+
+	// Resource group provided by user or a non default kubelet identity is used on the cluster
```

Comment:
We can get the cluster name from the pod spec
https://github.com/virtual-kubelet/azure-aci/pull/240/files#diff-f3b87d76a226c19ab2a6db3f4e284949c846b4746dd03f25c6f5f7430553d1aeR360